File: AlgorithmsProject/src/in/algorithms/balanced/BalancedExpression.scala
package in.algorithms.balanced

object BalancedExpression {

  def checkIfBalanced(givenString: String, currentCount: Int): Boolean = {
    if (givenString.isEmpty()) currentCount == 0
    else if (givenString.head == '(') checkIfBalanced(givenString.tail, currentCount + 1)
    else if ((givenString.head == ')')) currentCount > 0 && checkIfBalanced(givenString.tail, currentCount - 1)
    else checkIfBalanced(givenString.tail, currentCount)
  }
  
  def main(args: Array[String]) {
	  println("(a+b)-((a-b))*(b) is balanced :"+ checkIfBalanced("(a+b)-((a-b))*(b)", 0) )
	  println("a+b)-((a-b))*(b) is balanced :"+ checkIfBalanced("a+b)-((a-b))*(b)", 0) )
}
}

File: AlgorithmsProject/src/in/algorithms/balanced/BalancedExpression.java
package in.algorithms.balanced;

public class BalancedExpression {
    public static boolean checkIfBalanced(String s, int currentCount) {
        if (s.isEmpty()) return currentCount == 0;
        char head = s.charAt(0);
        String tail = s.substring(1);
        if (head == '(') return checkIfBalanced(tail, currentCount + 1);
        if (head == ')') return currentCount > 0 && checkIfBalanced(tail, currentCount - 1);
        return checkIfBalanced(tail, currentCount);
    }

    public static void main(String[] args) {
        System.out.println("(a+b)-((a-b))*(b) balanced: " + checkIfBalanced("(a+b)-((a-b))*(b)", 0));
        System.out.println("a+b)-((a-b))*(b) balanced: " + checkIfBalanced("a+b)-((a-b))*(b)", 0));
    }
}
